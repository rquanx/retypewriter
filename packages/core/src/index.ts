/**
 * 原理
 * 1、获取输入、输出 ==> 编辑前后的两块代码
 * 2、通过diff库，得到不变、插入、删除的代码块
 * 3、基于diff结果，计算出代码块变化的时候的位置即光标位置，删除的代码，是从哪个位置删除的（删除是从后往前删除），插入的时候是在那个位置插入的
 * 4、基于上述结果和input，创建generator,拆分分每次返回单个字符的变化记过，即每次增加/删除一个字符后，当前光标、字符、内容
 * 5、先清空编辑器，然后基于每次单个字符变化的信息，操作编辑器，设定光标、删除特定位置内容或插入字符
 */

export * from './type'
export * from './animator'
export * from './patch'
