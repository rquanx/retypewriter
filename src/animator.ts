import type { Patch } from './type'
export function* createAnimator(input: string, patches: Patch[]) {
  let output = input
  let cursor = 0
  for (const patch of patches) {
    if (patch.type === 'insert') {
      cursor = patch.from
      const head = output.slice(0, patch.from)
      const tail = output.slice(patch.from)

      let selection = ''
      for (const char of patch.text) {
        selection += char
        yield { cursor: cursor + selection.length, output: head + selection + tail }
      }
      // 插入的时候，拿到index前所有字符，拼接当前插入字符 加上剩下字符
      output = head + patch.text + tail
    }

    else if (patch.type === 'removal') {
      cursor = patch.from + patch.length
      const head = output.slice(0, patch.from - patch.length)
      const tail = output.slice(patch.from)
      const selection = output.slice(patch.from - patch.length, patch.from)
      for (let i = selection.length; i >= 0; i--)
        yield { cursor: cursor + 1, output: head + selection.slice(0, i) + tail }

      // 删除的时候，拿到当前index - 删除长度的字符，加上剩下字符
      output = head + tail
    }
  }
}
