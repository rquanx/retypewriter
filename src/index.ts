export * from './type'
export * from './animator'
export * from './patch'
