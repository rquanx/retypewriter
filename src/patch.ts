import type { Diff } from 'diff-match-patch'
import { diff_match_patch as DMP } from 'diff-match-patch'
import type { Patch } from './type'

export const diff = (a: string, b: string): Diff[] => {
  const differ = new DMP()
  const delta = differ.diff_main(a, b)
  differ.diff_cleanupSemantic(delta)
  return delta
}

export const calculatePatch = (diff: Diff[]): Patch[] => {
  const patches: Patch[] = []
  let index = 0
  for (const change of diff) {
    if (change[0] === 0) {
      index += change[1].length
      continue
    }
    if (change[0] === -1) {
      const length = change[1].length
      patches.push({
        type: 'removal',
        // 因为是从后往前删除，所以删除的位置是index + 删除的长度
        from: index + length,
        length,
      })
      // 删除后，index不变，即光标位置不变
    }
    if (change[0] === 1) {
      patches.push({
        type: 'insert',
        from: index,
        text: change[1],
      })
      index += change[1].length
    }
  }
  return patches
}
export const applyPatches = (input: string, patches: Patch[]) => {
  let output = input
  for (const patch of patches) {
    if (patch.type === 'insert')
      // 插入的时候，拿到index前所有字符，拼接当前插入字符 加上剩下字符
      output = output.slice(0, patch.from) + patch.text + output.slice(patch.from)

    else if (patch.type === 'removal')
      // 删除的时候，拿到当前index - 删除长度的字符，加上剩下字符
      output = output.slice(0, patch.from - patch.length) + output.slice(patch.from)
  }
  return { output }
}
