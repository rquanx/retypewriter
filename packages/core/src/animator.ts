import { calculatePatch, diff } from './patch'
import type { Patch } from './type'

interface AnimatorStep {
  cursor: number
  content: string
  patch: Patch
  patchIndex: number
  char?: string
}
export function* createAnimator(input: string, patches: Patch[]): Generator<AnimatorStep> {
  let output = input
  let cursor = 0
  for (let patchIndex = 0; patchIndex < patches.length; patchIndex++) {
    const patch = patches[patchIndex]
    if (patch.type === 'insert') {
      cursor = patch.from
      const head = output.slice(0, patch.from)
      const tail = output.slice(patch.from)

      let selection = ''
      for (const char of patch.text) {
        selection += char
        yield {
          char,
          cursor: cursor + selection.length,
          content: head + selection + tail,
          patch,
          patchIndex,
        }
      }
      // 插入的时候，拿到index前所有字符，拼接当前插入字符 加上剩下字符
      output = head + patch.text + tail
    }

    else if (patch.type === 'removal') {
      cursor = patch.from - patch.length // 获取最前的光标位置
      const head = output.slice(0, cursor)
      const tail = output.slice(patch.from)
      const selection = output.slice(cursor, patch.from) // 选中删除的内容
      for (let i = selection.length - 1; i >= 0; i--)
        yield { cursor: cursor + i, content: head + selection.slice(0, i) + tail, patch, patchIndex }

      // 删除的时候，拿到当前index - 删除长度的字符，加上剩下字符
      output = head + tail
    }
  }
}
export function simpleAnimator(input: string, output: string) {
  const delta = diff(input, output)
  const patches = calculatePatch(delta)
  return createAnimator(input, patches)
}
