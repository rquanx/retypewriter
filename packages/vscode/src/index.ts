import { Range, Selection, commands, window } from 'vscode'
import { simpleAnimator } from '../../core/src'
const json = [
  { content: '', time: 1648235120982 },
  { content: 'import { React } from \'react\'\n', time: 1648235136758 },
  { content: 'import { React } from \'react\'\nconst App = () => <div></div>\n', time: 1648235145833 },
  { content: 'import { React } from \'react\'\nconst App = () => <c></c>\n', time: 1648235149827 },
  { content: 'import { React } from \'react\'\nconst App = () => <div></div>\n', time: 1648235152018 }]
interface Snapshot {
  content: string
  time: number
}

const sleep = (ms: number) => {
  return new Promise(resolve => setTimeout(resolve, ms))
}
export function activate() {
  window.showInformationMessage('Hi')
  const snapMap = new Map<string, Snapshot[]>()

  commands.registerCommand('retypewriter.snap', () => {
    const doc = window.activeTextEditor?.document
    if (!doc) return
    const path = doc.uri.fsPath
    if (!snapMap.has(path))
      snapMap.set(path, [])
    const snaps = snapMap.get(path)
    snaps!.push({
      content: doc.getText(),
      time: Date.now(),
    })
    window.showInformationMessage(`reTypewriter: snapshot added (${snaps?.length})`)
  })
  commands.registerCommand('retypewriter.play', async() => {
    const editor = window.activeTextEditor
    const doc = editor?.document
    if (!doc || !editor) return
    const path = doc.uri.fsPath
    const snaps = snapMap.get(path)
    if (!snaps?.length) {
      window.showInformationMessage('reTypewriter: No snapshots found')
      return
    }
    const lastSnap = snaps[snaps.length - 1]
    if (lastSnap.content !== doc.getText()) {
      snaps.push({
        content: doc.getText(),
        time: Date.now(),
      })
    }

    window.showInformationMessage('reTypewriter: reTypewriter: Playing...')

    let lastContent: string | undefined
    for (const snap of snaps) {
      if (!lastContent) {
        lastContent = snap.content
        // clear doc
        await editor.edit((edit) => {
          edit.replace(new Range(0, 0, Infinity, Infinity), lastContent!)
        })
        continue
      }
      const animator = simpleAnimator(lastContent, snap.content)

      let lastIndex = -1
      for (const result of animator) {
        if (lastIndex !== result.patchIndex)
          await sleep(900)
        const pos = doc.positionAt(result.cursor)
        await editor.edit((edit) => {
          if (result.char != null)
            edit.insert(doc.positionAt(result.cursor - 1), result.char)
          else
            edit.delete(new Range(pos, doc.positionAt(result.cursor + 1)))
        })
        editor.selection = new Selection(pos, pos)
        await sleep(Math.random() * 200)
        lastIndex = result.patchIndex
      }
      lastContent = snap.content
    }
    window.showInformationMessage('reTypewriter: reTypewriter: Finished...')
  })
}

export function deactivate() {

}
