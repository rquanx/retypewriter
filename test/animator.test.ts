import { expect, it } from 'vitest'
import { applyPatches, calculatePatch, createAnimator, diff } from '../src'

import { input, output } from './fixture'
it('animator', () => {
  const delta = diff(input, output)
  const patches = calculatePatch(delta)
  const applied = applyPatches(input, patches)
  const animator = createAnimator(applied.output, patches)
  expect([...animator]).toMatchSnapshot('animator')
})
