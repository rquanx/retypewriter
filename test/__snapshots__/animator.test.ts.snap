// Vitest Snapshot v1

exports[`animator > animator 1`] = `
[
  {
    "cursor": 105,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 105,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 105,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () =>   it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 105,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () =>  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 105,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () =  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 105,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', ()   it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 105,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', ()  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 105,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', (  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 105,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should',   it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 105,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should',  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 105,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should'  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 105,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 105,
    "output": "
import { describe, expect, it } from 'vitest'

describe('shoul  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 105,
    "output": "
import { describe, expect, it } from 'vitest'

describe('shou  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 105,
    "output": "
import { describe, expect, it } from 'vitest'

describe('sho  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 105,
    "output": "
import { describe, expect, it } from 'vitest'

describe('sh  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 105,
    "output": "
import { describe, expect, it } from 'vitest'

describe('s  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 105,
    "output": "
import { describe, expect, it } from 'vitest'

describe('  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 105,
    "output": "
import { describe, expect, it } from 'vitest'

describe(  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 105,
    "output": "
import { describe, expect, it } from 'vitest'

describe  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 105,
    "output": "
import { describe, expect, it } from 'vitest'

describ  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 105,
    "output": "
import { describe, expect, it } from 'vitest'

descri  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 105,
    "output": "
import { describe, expect, it } from 'vitest'

descr  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 105,
    "output": "
import { describe, expect, it } from 'vitest'

desc  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 105,
    "output": "
import { describe, expect, it } from 'vitest'

des  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 105,
    "output": "
import { describe, expect, it } from 'vitest'

de  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 105,
    "output": "
import { describe, expect, it } from 'vitest'

d  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 105,
    "output": "
import { describe, expect, it } from 'vitest'

  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 105,
    "output": "
import { describe, expect, it } from 'vitest'
  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 105,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 98,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one).toEqual(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 98,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one).toEqua(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 98,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one).toEqu(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 98,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one).toEq(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 98,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one).toE(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 98,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one).to(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 98,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one).t(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 98,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one).(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 98,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one)(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 82,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one)o(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 83,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one)on(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 84,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one)one(1)
    expect(two).toEqual(2)
  })
})
",
  },
  {
    "cursor": 123,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one)one(1)
    expect(two).toEqual(2)
  })
}) 
",
  },
  {
    "cursor": 124,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one)one(1)
    expect(two).toEqual(2)
  })
})  
",
  },
  {
    "cursor": 125,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one)one(1)
    expect(two).toEqual(2)
  })
})   
",
  },
  {
    "cursor": 126,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one)one(1)
    expect(two).toEqual(2)
  })
})    
",
  },
  {
    "cursor": 127,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one)one(1)
    expect(two).toEqual(2)
  })
})    e
",
  },
  {
    "cursor": 128,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one)one(1)
    expect(two).toEqual(2)
  })
})    ex
",
  },
  {
    "cursor": 129,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one)one(1)
    expect(two).toEqual(2)
  })
})    exp
",
  },
  {
    "cursor": 130,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one)one(1)
    expect(two).toEqual(2)
  })
})    expe
",
  },
  {
    "cursor": 131,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one)one(1)
    expect(two).toEqual(2)
  })
})    expec
",
  },
  {
    "cursor": 132,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one)one(1)
    expect(two).toEqual(2)
  })
})    expect
",
  },
  {
    "cursor": 133,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one)one(1)
    expect(two).toEqual(2)
  })
})    expect(
",
  },
  {
    "cursor": 134,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one)one(1)
    expect(two).toEqual(2)
  })
})    expect(t
",
  },
  {
    "cursor": 135,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one)one(1)
    expect(two).toEqual(2)
  })
})    expect(tw
",
  },
  {
    "cursor": 136,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one)one(1)
    expect(two).toEqual(2)
  })
})    expect(two
",
  },
  {
    "cursor": 137,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one)one(1)
    expect(two).toEqual(2)
  })
})    expect(two)
",
  },
  {
    "cursor": 138,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one)one(1)
    expect(two).toEqual(2)
  })
})    expect(two).
",
  },
  {
    "cursor": 139,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one)one(1)
    expect(two).toEqual(2)
  })
})    expect(two).t
",
  },
  {
    "cursor": 140,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one)one(1)
    expect(two).toEqual(2)
  })
})    expect(two).to
",
  },
  {
    "cursor": 141,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one)one(1)
    expect(two).toEqual(2)
  })
})    expect(two).toE
",
  },
  {
    "cursor": 142,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one)one(1)
    expect(two).toEqual(2)
  })
})    expect(two).toEq
",
  },
  {
    "cursor": 143,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one)one(1)
    expect(two).toEqual(2)
  })
})    expect(two).toEqu
",
  },
  {
    "cursor": 144,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one)one(1)
    expect(two).toEqual(2)
  })
})    expect(two).toEqua
",
  },
  {
    "cursor": 145,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one)one(1)
    expect(two).toEqual(2)
  })
})    expect(two).toEqual
",
  },
  {
    "cursor": 146,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one)one(1)
    expect(two).toEqual(2)
  })
})    expect(two).toEqual(
",
  },
  {
    "cursor": 147,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one)one(1)
    expect(two).toEqual(2)
  })
})    expect(two).toEqual(2
",
  },
  {
    "cursor": 148,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one)one(1)
    expect(two).toEqual(2)
  })
})    expect(two).toEqual(2)
",
  },
  {
    "cursor": 149,
    "output": "
import { describe, expect, it } from 'vitest'  it('one', () => {
    expect(one)one(1)
    expect(two).toEqual(2)
  })
})    expect(two).toEqual(2)

",
  },
]
`;
